File: mailblastr-php/tests/cases/ResourcesTest.php
<?php

declare(strict_types=1);

// ---- audiences (incl. importSheet) ----
[$mb, $t] = make_client();
$mb->audiences->create(['name' => 'General']);
check_same('audiences.create: path', '/audiences', $t->lastPath());

$mb->audiences->importSheet('aud_1', ['url' => 'https://docs.google.com/spreadsheets/d/x', 'segment_name' => 'Sheet import']);
check_same('audiences.importSheet: path', '/audiences/aud_1/contacts/import-sheet', $t->lastPath());
check_same('audiences.importSheet: body', ['url' => 'https://docs.google.com/spreadsheets/d/x', 'segment_name' => 'Sheet import'], $t->lastJson());

$mb->audiences->update('aud_1', ['name' => 'Renamed']);
check_same('audiences.update: method', 'PATCH', $t->last()['method']);
$mb->audiences->remove('aud_1');
check_same('audiences.remove: method', 'DELETE', $t->last()['method']);

// ---- contact properties ----
$mb->contactProperties->create(['key' => 'plan', 'type' => 'string', 'fallback_value' => 'free']);
check_same('contactProperties.create: path', '/contact-properties', $t->lastPath());
$mb->contactProperties->list(['limit' => 10]);
check_same('contactProperties.list: path', '/contact-properties?limit=10', $t->lastPath());
$mb->contactProperties->update('prop_1', ['fallback_value' => 'pro']);
check_same('contactProperties.update: path', '/contact-properties/prop_1', $t->lastPath());
$mb->contactProperties->remove('prop_1');
check_same('contactProperties.remove: method', 'DELETE', $t->last()['method']);

// ---- segments (domain-required) ----
$mb->segments->create(['domain' => 'yourdomain.com', 'name' => 'VIP', 'filter' => ['status' => 'subscribed']]);
check_same('segments.create: path', '/segments', $t->lastPath());
check_same('segments.create: domain in body', ['domain' => 'yourdomain.com', 'name' => 'VIP', 'filter' => ['status' => 'subscribed']], $t->lastJson());

$mb->segments->list(['domain' => 'yourdomain.com', 'limit' => 5]);
check_same('segments.list: domain query required', '/segments?domain=yourdomain.com&limit=5', $t->lastPath());

$mb->segments->contacts('seg_1');
check_same('segments.contacts: path', '/segments/seg_1/contacts', $t->lastPath());
$mb->segments->contacts('seg_1', ['limit' => 50]);
check_same('segments.contacts: pagination', '/segments/seg_1/contacts?limit=50', $t->lastPath());
$mb->segments->update('seg_1', ['name' => 'VIP+']);
check_same('segments.update: method', 'PATCH', $t->last()['method']);
$mb->segments->remove('seg_1');
check_same('segments.remove: method', 'DELETE', $t->last()['method']);

// ---- topics (domain-required) ----
$mb->topics->create(['domain' => 'yourdomain.com', 'name' => 'Product updates', 'default_subscription' => 'opt_in']);
check_same('topics.create: path', '/topics', $t->lastPath());
check_same('topics.create: domain in body', ['domain' => 'yourdomain.com', 'name' => 'Product updates', 'default_subscription' => 'opt_in'], $t->lastJson());
$mb->topics->list(['domain' => 'yourdomain.com']);
check_same('topics.list: domain query', '/topics?domain=yourdomain.com', $t->lastPath());
$mb->topics->update('top_1', ['visibility' => 'private']);
check_same('topics.update: path', '/topics/top_1', $t->lastPath());
$mb->topics->remove('top_1');
check_same('topics.remove: method', 'DELETE', $t->last()['method']);

// ---- campaigns (domain-required; empty send payload → {}) ----
$mb->campaigns->create(['domain' => 'yourdomain.com', 'from' => 'a@yourdomain.com', 'subject' => 'S', 'html' => '<p>x</p>', 'segment_id' => 'seg_1']);
check_same('campaigns.create: path', '/campaigns', $t->lastPath());
check_same('campaigns.create: domain in body', 'yourdomain.com', $t->lastJson()['domain'] ?? null);

$mb->campaigns->send('cmp_1');
check_same('campaigns.send: path', '/campaigns/cmp_1/send', $t->lastPath());
check_same('campaigns.send: empty payload serializes as {}', '{}', $t->last()['body']);

$mb->campaigns->send('cmp_1', ['scheduled_at' => '2026-08-01T00:00:00Z']);
check_same('campaigns.send: scheduled body', ['scheduled_at' => '2026-08-01T00:00:00Z'], $t->lastJson());

$mb->campaigns->cancel('cmp_1');
check_same('campaigns.cancel: path', '/campaigns/cmp_1/cancel', $t->lastPath());
$mb->campaigns->stats('cmp_1');
check_same('campaigns.stats: path', '/campaigns/cmp_1/stats', $t->lastPath());
$mb->campaigns->engagement('cmp_1');
check_same('campaigns.engagement: method', 'GET', $t->last()['method']);
check_same('campaigns.engagement: path', '/campaigns/cmp_1/engagement', $t->lastPath());
$mb->campaigns->ab('cmp_1');
check_same('campaigns.ab: path', '/campaigns/cmp_1/ab', $t->lastPath());
$mb->campaigns->remove('cmp_1');
check_same('campaigns.remove: method', 'DELETE', $t->last()['method']);

// ---- templates ----
$mb->templates->create(['name' => 'Welcome', 'subject' => 'Hi {{first_name}}', 'html' => '<p>Hi</p>']);
check_same('templates.create: path', '/templates', $t->lastPath());
$mb->templates->duplicate('tmpl_1');
check_same('templates.duplicate: path', '/templates/tmpl_1/duplicate', $t->lastPath());
check_same('templates.duplicate: empty payload serializes as {}', '{}', $t->last()['body']);
$mb->templates->publish('tmpl_1');
check_same('templates.publish: path', '/templates/tmpl_1/publish', $t->lastPath());
check_same('templates.publish: no body', null, $t->last()['body']);
$mb->templates->update('tmpl_1', ['name' => 'Welcome v2']);
check_same('templates.update: method', 'PATCH', $t->last()['method']);
$mb->templates->remove('tmpl_1');
check_same('templates.remove: method', 'DELETE', $t->last()['method']);

// ---- automations (domain REQUIRED on create) ----
$mb->automations->create(['name' => 'Welcome series', 'domain' => 'yourdomain.com', 'trigger' => 'contact.created']);
check_same('automations.create: path', '/automations', $t->lastPath());
check_same('automations.create: body', ['name' => 'Welcome series', 'domain' => 'yourdomain.com', 'trigger' => 'contact.created'], $t->lastJson());

$mb->automations->addStep('auto_1', ['type' => 'send_email', 'config' => ['template_id' => 'tmpl_welcome']]);
check_same('automations.addStep: path', '/automations/auto_1/steps', $t->lastPath());

// A step update REPLACES the step, it does not merge into it: the API re-reads
// the whole body and rejects a missing 'type' with a 422 validation_error
// BEFORE it ever inspects 'config'. So a type-less PATCH is not "leave the type
// alone", it is an error every time — the key must be on the wire.
$mb->automations->updateStep('auto_1', 'step_2', ['type' => 'delay', 'config' => ['duration' => '3 days']]);
check_same('automations.updateStep: method', 'PATCH', $t->last()['method']);
check_same('automations.updateStep: path', '/automations/auto_1/steps/step_2', $t->lastPath());
check('automations.updateStep: type on the wire (a type-less body is a 422)', array_key_exists('type', $t->lastJson()));
check_same('automations.updateStep: body', ['type' => 'delay', 'config' => ['duration' => '3 days']], $t->lastJson());

$mb->automations->deleteStep('auto_1', 'step_2');
check_same('automations.deleteStep: method', 'DELETE', $t->last()['method']);
check_same('automations.deleteStep: path', '/automations/auto_1/steps/step_2', $t->lastPath());

$mb->automations->createWithAi('auto_1', ['prompt' => 'Welcome new signups over 3 days']);
check_same('automations.createWithAi: method', 'POST', $t->last()['method']);
check_same('automations.createWithAi: path', '/automations/auto_1/ai', $t->lastPath());
check_same('automations.createWithAi: body', ['prompt' => 'Welcome new signups over 3 days'], $t->lastJson());

$mb->automations->runs('auto_1', ['limit' => 25]);
check_same('automations.runs: path', '/automations/auto_1/runs?limit=25', $t->lastPath());

$mb->automations->runs('auto_1', ['status' => 'failed,running']);
check_same('automations.runs: status filter', '/automations/auto_1/runs?status=failed%2Crunning', $t->lastPath());

$mb->automations->getRun('auto_1', 'run_1');
check_same('automations.getRun: path', '/automations/auto_1/runs/run_1', $t->lastPath());

$mb->automations->stop('auto_1');
check_same('automations.stop: path', '/automations/auto_1/stop', $t->lastPath());
$mb->automations->update('auto_1', ['status' => 'enabled']);
check_same('automations.update: method', 'PATCH', $t->last()['method']);
$mb->automations->remove('auto_1');
check_same('automations.remove: method', 'DELETE', $t->last()['method']);

// ---- webhooks (HTTP endpoints; verify() is covered in WebhookVerifyTest) ----
$mb->webhooks->create(['endpoint' => 'https://yourapp.com/hooks', 'events' => ['email.delivered']]);
check_same('webhooks.create: path', '/webhooks', $t->lastPath());
$mb->webhooks->rotate('wh_1');
check_same('webhooks.rotate: path', '/webhooks/wh_1/rotate', $t->lastPath());
$mb->webhooks->test('wh_1');
check_same('webhooks.test: path', '/webhooks/wh_1/test', $t->lastPath());
// A FAILED test delivery is still HTTP 200 — it must not throw, and the real
// outcome is 'ok', never the status code.
$t->queue(200, ['object' => 'webhook_test', 'id' => 'wh_1', 'ok' => false, 'error' => 'lookup_failed']);
$failed = $mb->webhooks->test('wh_1');
check_same('webhooks.test: failed delivery reports ok=false', false, $failed['ok']);
check_same('webhooks.test: failed delivery names the reason', 'lookup_failed', $failed['error']);
check('webhooks.test: failed delivery carries no endpoint status', !isset($failed['status']));
$t->queue(200, ['object' => 'webhook_test', 'id' => 'wh_1', 'ok' => true, 'status' => 200]);
$delivered = $mb->webhooks->test('wh_1');
check_same('webhooks.test: successful delivery reports ok=true', true, $delivered['ok']);
check_same('webhooks.test: successful delivery carries the endpoint status', 200, $delivered['status']);
$mb->webhooks->update('wh_1', ['status' => 'disabled']);
check_same('webhooks.update: method', 'PATCH', $t->last()['method']);
$mb->webhooks->remove('wh_1');
check_same('webhooks.remove: method', 'DELETE', $t->last()['method']);

// ---- events (domain REQUIRED on send) ----
$mb->events->send(['event' => 'signup.completed', 'domain' => 'yourdomain.com', 'email' => 'user@example.com', 'payload' => ['plan' => 'pro']]);
check_same('events.send: path', '/events/send', $t->lastPath());
check_same('events.send: body', ['event' => 'signup.completed', 'domain' => 'yourdomain.com', 'email' => 'user@example.com', 'payload' => ['plan' => 'pro']], $t->lastJson());
// No options => no header at all.
check('events.send: no idempotency header by default', !array_filter(
    $t->last()['headers'],
    static fn (string $h): bool => str_starts_with($h, 'Idempotency-Key:')
));

// The $options parameter is still part of the signature (it is shared with
// every other resource) and a key given here is still forwarded — but only
// /emails and /emails/batch HONOUR the header, so this buys nothing. The
// docblock says so; the parameter stays so existing call sites keep working.
$mb->events->send(
    ['event' => 'signup.completed', 'domain' => 'yourdomain.com', 'email' => 'user@example.com'],
    ['idempotencyKey' => 'evt-1']
);
check('events.send: accepts and forwards an idempotency key', in_array('Idempotency-Key: evt-1', $t->last()['headers'], true));

$mb->events->create(['name' => 'signup.completed', 'schema' => ['plan' => 'string']]);
check_same('events.create: path', '/events', $t->lastPath());
$mb->events->create(['name' => 'signup.completed'], ['idempotencyKey' => 'evt-2']);
check('events.create: accepts and forwards an idempotency key', in_array('Idempotency-Key: evt-2', $t->last()['headers'], true));
$mb->events->list(['limit' => 10]);
check_same('events.list: path', '/events?limit=10', $t->lastPath());
$mb->events->update('evt_1', ['schema' => ['plan' => 'string']]);
check_same('events.update: method', 'PATCH', $t->last()['method']);
check_same('events.update: path', '/events/evt_1', $t->lastPath());
check_same('events.update: body', ['schema' => ['plan' => 'string']], $t->lastJson());
$mb->events->remove('evt_1');
check_same('events.remove: path', '/events/evt_1', $t->lastPath());

// ---- api keys (listing only; lifecycle is dashboard-only) ----
$mb->apiKeys->list();
check_same('apiKeys.list: path', '/api-keys', $t->lastPath());
$mb->apiKeys->list(['limit' => 5]);
check_same('apiKeys.list: pagination', '/api-keys?limit=5', $t->lastPath());
foreach (['create', 'update', 'remove', 'delete', 'revoke'] as $absent) {
    check(
        "apiKeys.$absent: absent (key lifecycle is dashboard-only)",
        !method_exists($mb->apiKeys, $absent)
    );
}

// ---- logs (method/status filters) ----
$mb->logs->list(['limit' => 100, 'method' => 'POST', 'status' => 429]);
check_same('logs.list: filters in query', '/logs?limit=100&method=POST&status=429', $t->lastPath());
$mb->logs->get('log_1');
check_same('logs.get: path', '/logs/log_1', $t->lastPath());

// ---- polls ----
$mb->polls->list(['limit' => 10]);
check_same('polls.list: path', '/polls?limit=10', $t->lastPath());
$mb->polls->get('em_1');
check_same('polls.get: path', '/polls/em_1', $t->lastPath());

File: mailblastr-ruby/test/resources_test.rb
# frozen_string_literal: true

require "test_helper"

# Route-level sweep of the remaining resources: audiences, contact properties,
# topics, templates, automations, events, api keys, logs, polls.
class ResourcesTest < Minitest::Test
  include ClientStubHelper

  def test_audiences
    Mailblastr::Audiences.create({ name: "Newsletter" })
    assert_request :post, "/audiences"

    Mailblastr::Audiences.get("aud_1")
    assert_request :get, "/audiences/aud_1"

    Mailblastr::Audiences.list({ limit: 3 })
    assert_request :get, "/audiences?limit=3"

    Mailblastr::Audiences.update("aud_1", { name: "Weekly" })
    assert_request :patch, "/audiences/aud_1"

    Mailblastr::Audiences.import_sheet("aud_1", { url: "https://docs.google.com/x", segment_name: "June" })
    assert_request :post, "/audiences/aud_1/contacts/import-sheet"
    assert_equal "June", last_body["segment_name"]

    Mailblastr::Audiences.delete("aud_1")
    assert_request :delete, "/audiences/aud_1"
  end

  def test_contact_properties
    Mailblastr::ContactProperties.create({ key: "plan", type: "string", fallback_value: "free" })
    assert_request :post, "/contact-properties"
    assert_equal "plan", last_body["key"]

    Mailblastr::ContactProperties.get("prop_1")
    assert_request :get, "/contact-properties/prop_1"

    Mailblastr::ContactProperties.list
    assert_request :get, "/contact-properties"

    Mailblastr::ContactProperties.update("prop_1", { fallback_value: "pro" })
    assert_request :patch, "/contact-properties/prop_1"

    Mailblastr::ContactProperties.delete("prop_1")
    assert_request :delete, "/contact-properties/prop_1"
  end

  def test_topics_domain_required
    assert_raises(ArgumentError) { Mailblastr::Topics.create({ name: "Updates" }) }
    assert_raises(ArgumentError) { Mailblastr::Topics.list({}) }

    Mailblastr::Topics.create({ domain: "yourdomain.com", name: "Updates", default_subscription: "opt_in" })
    assert_request :post, "/topics"
    assert_equal "yourdomain.com", last_body["domain"]

    Mailblastr::Topics.list({ domain: "yourdomain.com", limit: 50 })
    assert_request :get, "/topics?domain=yourdomain.com&limit=50"

    Mailblastr::Topics.get("top_1")
    assert_request :get, "/topics/top_1"

    Mailblastr::Topics.update("top_1", { description: "News" })
    assert_request :patch, "/topics/top_1"

    Mailblastr::Topics.delete("top_1")
    assert_request :delete, "/topics/top_1"
  end

  def test_templates
    Mailblastr::Templates.create({ name: "Welcome", subject: "Hi", html: "<p>Hi</p>" })
    assert_request :post, "/templates"

    Mailblastr::Templates.get("tmpl_1")
    assert_request :get, "/templates/tmpl_1"

    Mailblastr::Templates.list
    assert_request :get, "/templates"

    Mailblastr::Templates.update("tmpl_1", { subject: "Hello" })
    assert_request :patch, "/templates/tmpl_1"

    Mailblastr::Templates.duplicate("tmpl_1", { name: "Welcome v2" })
    assert_request :post, "/templates/tmpl_1/duplicate"
    assert_equal "Welcome v2", last_body["name"]

    Mailblastr::Templates.duplicate("tmpl_1")
    assert_equal({}, last_body)

    Mailblastr::Templates.publish("tmpl_1")
    assert_request :post, "/templates/tmpl_1/publish"

    Mailblastr::Templates.delete("tmpl_1")
    assert_request :delete, "/templates/tmpl_1"
  end

  def test_automations_domain_required_and_full_surface
    assert_raises(ArgumentError) { Mailblastr::Automations.create({ name: "Welcome" }) }

    Mailblastr::Automations.create({ name: "Welcome", domain: "yourdomain.com", trigger: "contact.created" })
    assert_request :post, "/automations"
    assert_equal "yourdomain.com", last_body["domain"]

    Mailblastr::Automations.get("auto_1")
    assert_request :get, "/automations/auto_1"

    Mailblastr::Automations.list({ limit: 10 })
    assert_request :get, "/automations?limit=10"

    Mailblastr::Automations.update("auto_1", { status: "enabled" })
    assert_request :patch, "/automations/auto_1"

    Mailblastr::Automations.add_step("auto_1", { type: "send_email", config: { template_id: "tmpl_1" } })
    assert_request :post, "/automations/auto_1/steps"
    assert_equal "send_email", last_body["type"]

    # `type` must ride along: the server re-validates the whole step, so a
    # type-less body is a 422 no matter how valid the config is.
    Mailblastr::Automations.update_step("auto_1", "step_1",
                                        { type: "send_email", config: { template_id: "tmpl_1", subject: "Hi" } })
    assert_request :patch, "/automations/auto_1/steps/step_1"
    assert_equal "send_email", last_body["type"]
    assert_equal({ "template_id" => "tmpl_1", "subject" => "Hi" }, last_body["config"])

    Mailblastr::Automations.delete_step("auto_1", "step_1")
    assert_request :delete, "/automations/auto_1/steps/step_1"

    Mailblastr::Automations.create_with_ai("auto_1", { prompt: "Wait 2 days then send the welcome email" })
    assert_request :post, "/automations/auto_1/ai"
    assert_equal "Wait 2 days then send the welcome email", last_body["prompt"]

    Mailblastr::Automations.runs("auto_1", { limit: 25 })
    assert_request :get, "/automations/auto_1/runs?limit=25"

    Mailblastr::Automations.get_run("auto_1", "run_1")
    assert_request :get, "/automations/auto_1/runs/run_1"

    Mailblastr::Automations.stop("auto_1")
    assert_request :post, "/automations/auto_1/stop"

    Mailblastr::Automations.delete("auto_1")
    assert_request :delete, "/automations/auto_1"
  end

  def test_events_domain_required_on_send
    assert_raises(ArgumentError) { Mailblastr::Events.send({ event: "signup.completed", email: "a@b.com" }) }

    Mailblastr::Events.send(
      { event: "signup.completed", domain: "yourdomain.com", email: "a@b.com", payload: { plan: "pro" } },
      { idempotency_key: "evt-1" }
    )
    assert_request :post, "/events/send"
    assert_equal "yourdomain.com", last_body["domain"]
    assert_equal "evt-1", last_request["Idempotency-Key"]

    Mailblastr::Events.create({ name: "signup.completed", schema: { plan: "string" } })
    assert_request :post, "/events"

    Mailblastr::Events.list
    assert_request :get, "/events"

    Mailblastr::Events.update("evt_1", { schema: { plan: "string", seats: "number" } })
    assert_request :patch, "/events/evt_1"
    assert_equal({ "plan" => "string", "seats" => "number" }, last_body["schema"])

    Mailblastr::Events.delete("evt_1")
    assert_request :delete, "/events/evt_1"
  end

  def test_api_keys
    Mailblastr::ApiKeys.list
    assert_request :get, "/api-keys"

    Mailblastr::ApiKeys.list({ limit: 10 })
    assert_request :get, "/api-keys?limit=10"
  end

  # Key lifecycle is dashboard-only, so the SDK exposes nothing that would
  # reach POST /api-keys, PATCH /api-keys/:id or DELETE /api-keys/:id.
  def test_api_key_lifecycle_methods_are_absent
    %i[create update delete remove revoke].each do |name|
      refute_respond_to Mailblastr::ApiKeys, name, "Mailblastr::ApiKeys.#{name} must not exist"
    end
  end

  def test_logs_with_filters
    Mailblastr::Logs.list({ limit: 100, method: "POST", status: 429 })
    assert_request :get, "/logs?limit=100&method=POST&status=429"

    Mailblastr::Logs.get("log_1")
    assert_request :get, "/logs/log_1"
  end

  def test_automation_runs_status_filter_accepts_a_string_or_an_array
    Mailblastr::Automations.runs("auto_1", { status: "failed" })
    assert_request :get, "/automations/auto_1/runs?status=failed"

    # The server takes a comma-separated list; an Array is joined for you.
    Mailblastr::Automations.runs("auto_1", { limit: 5, status: %w[failed skipped] })
    assert_request :get, "/automations/auto_1/runs?limit=5&status=failed%2Cskipped"
  end

  def test_polls
    Mailblastr::Polls.list({ limit: 10 })
    assert_request :get, "/polls?limit=10"

    Mailblastr::Polls.get("email_1")
    assert_request :get, "/polls/email_1"
  end
end
